### ISSUE TITLE
ENH: stats.boxcox/yeojohnson: additional improvements w.r.t. overflow

### ISSUE THREAD

**Issue by mdhaber** [scipy/scipy#19016]
gh-18852 made several improvements to avoid overflow when computing the optimal Yeo Johnson transformation parameter - or, more specifically, to produce a parameter value that avoids overflow when the data is transformed. There were a few additional improvement ideas that should be either be implemented or rejected. 

- [ ] Adjust the treatment of mixed positive and negative data to avoid https://github.com/scipy/scipy/pull/18852#discussion_r1278616032.
- [ ] Inform the user when the provided parameter is not the true, unconstrained optimum but is instead a value that is optimal under a constraint that it avoids overflow when used to transform data.
- [ ] When finding the constrained optimum, invert the transformation exactly rather than using an approximation. (Discussed [here](https://github.com/scipy/scipy/pull/18852#pullrequestreview-1531164724).)
- [ ] Reconsider adjustments to the bounds of the constrained problem. Specifically, the machine precision adjustments (explained [here](https://github.com/scipy/scipy/pull/18852#issuecomment-1635511232)) may be too conservative. Also, it might be worth allowing the user to adjust the fudge factor (currently `20`) depending on their knowledge of their use case. (It also might not be important enough to add a parameter to the API.)
- [ ] Give the user the option to compute the true, unconstrained optimum by [performing the calculation in the log-space](https://github.com/scipy/scipy/pull/18852#issuecomment-1657858886).

gh-19604 made similar improvements to `boxcox` to avoid overflow, but there is still follow-up work.

- [x] Calculate the true, unconstrained optimum by [performing the calculation in the log-space](https://github.com/scipy/scipy/pull/18852#issuecomment-1657858886).
- [x] Calculate the constrained optimum, inverting the transformation exactly rather than using an approximation.
- [x] Inform the user when the provided parameter is not the true, unconstrained optimum but is instead a value that is optimal under a constraint that it avoids overflow when used to transform data.
- [x] Allow the user to control the maximum value of the transformed data. When the maximum is `np.inf`, return the true, unconstrained optimum.
- [x] Avoid premature overflow in `special.boxcox` itself. https://github.com/scipy/scipy/pull/19604#issuecomment-1833218049
- [ ] Investigate the feasibility of similar improvements with `method='pearson'`.

**Comment by lorentzenchr** [comment]
Thinking about handling overflow in box-cox once more, maybe it is enough what we have right now and add more if users really need it - as opposed to adding yet another parameter `xmax` which has its appeal.

**Comment by mdhaber** [comment]
@xuefeng-xu @lsorber would you be ok with that (https://github.com/scipy/scipy/issues/19016#issuecomment-1887982497)? Are there other remaining changes to be made to BC to set the precedent for YJ?

(Separately, would someone be willing to make a PR to fix the [premature overflow](https://github.com/scipy/scipy/pull/19604#issuecomment-1833218049) in `special.boxcox` itself?)

**Comment by xuefeng-xu** [comment]
I'm ok with that. May be we can finish the premature overflow issue in `special.boxcox` then go to YJ.

Here is what I thought for the premature overflow issue. 
1. Adjust the computation as the following, since `lmbda` might be negative.
```python
def boxcox(x, lmbda):
    return np.sign(lmbda) * np.exp(lmbda * np.log(x) - np.log(abs(lmbda))) - 1 / lmbda
```
2. Think about if the above change affects the line 12.
https://github.com/scipy/scipy/blob/4b0363716ef5a8229f6abddaf206fea318a2d6b7/scipy/special/_boxcox.pxd#L5-L13
3. Do the same thing for `special.boxcox1p`

**Comment by mdhaber** [comment]
1. Yes, that's essentially what I suggested with the addition of support for negative lambda. Looks good.

2. I'm not sure if it really needs to adjust line 12. TBH, one idea is to just try this version of calculation if the result of the original calculation overflows. I would be comfortable reviewing that because it is almost certain to do no harm. I'd suggest starting with that. 
If you really want to dive in, I would do some testing against the results you get with `mpmath` to study how different formulations perform across the $x$, $\lambda$ space. This would take longer to get reviewed, though; you'd need someone more familiar with `special` to take a look. Maybe @steppi would be interested.

4. ✓

**Comment by lsorber** [comment]
> @xuefeng-xu @lsorber would you be ok with that (https://github.com/scipy/scipy/issues/19016#issuecomment-1887982497)?

Yes, I'm OK with the solution not being `xmax`, but no I do not think that overflow prevention is fully addressed yet and that we should leave it as is. In its current state, overflow is prevented for training, but not for inference.

[As I show with an example in this comment](https://github.com/scipy/scipy/pull/19631#discussion_r1441470365), if the overflow prevention is triggered on training data `x`, overflow is almost certain to occur when applying the Box-Cox transform to data that is larger than `x`. And that's exactly what we should expect to happen with users of [sklearn.preprocessing.PowerTransformer](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.PowerTransformer.html).

Which solutions can we pick from to address overflow prevention during inference?
1. Introduce an `xmax`. Explicit but introduces yet another parameter.
2. Reduce the default `ymax`, [which currently has a 'buffer' of 4 digits](https://github.com/scipy/scipy/blob/af8374b97c349e55c82087233b3137311728859b/scipy/stats/_morestats.py#L1258C38-L1258C43) w.r.t. the largest floating point value. This is probably too much for float16, and it's unlikely to be sufficient to prevent overflow for inference with float64.
3. Assume a default value for `xmax = 20 * np.max(x)`.

My recommendation: I would reduce the buffer on `ymax` to the minimum necessary, introduce `xmax` as an explicit parameter, and assume a default value of `xmax = 20 * np.max(x)`.

**Comment by mdhaber** [comment]
Sure. In that case, I think just @xuefeng-xu and @lorentzenchr would need to be on board.

**Comment by xuefeng-xu** [comment]
I'm ok to reduce the buffer on ymax if we think it's necessacy.

**Comment by lsorber** [comment]
Concrete suggestion:

1. Change this to `ymax = np.finfo(dtype).max / 100` (this is to use the available range more efficiently):
https://github.com/scipy/scipy/blob/af8374b97c349e55c82087233b3137311728859b/scipy/stats/_morestats.py#L1258
2. Change this to `mask = abs(special.boxcox(100 * x_treme, res)) > ymax` (here, `xmax := 100 * x_treme` prevents overflow for x larger than the training data):
https://github.com/scipy/scipy/blob/af8374b97c349e55c82087233b3137311728859b/scipy/stats/_morestats.py#L1352

**Comment by mdhaber** [comment]
@lsorber sounded like @xuefeng-xu and @lorentzenchr were on board with increasing the buffer. Want to open a PR? I'll review after they do.
